packages/ra-ui-materialui/src/list/Pagination.js

```diff
@@ -25,16 +25,21 @@ const Pagination = ({
     limit,
     ...rest
 }) => {
+    const getNbPages = useCallback(() => Math.ceil(total / perPage) || 1, [
+        perPage,
+        total,
+    ]);
+
     useEffect(() => {
         if (page < 1 || isNaN(page)) {
             setPage(1);
+        } else if (page > getNbPages()) {
+            setPage(getNbPages());
         }
-    }, [page, setPage]);
+    }, [page, setPage, getNbPages, total, perPage]);
```

Comment:
no need for `perPage` or `total` as it's already in the `useCallback` dependency array above.